### PR TITLE
SONARPY-2080 Move ShannonEntropy to analyzer commons

### PR DESCRIPTION
**Pull request by ghislainpiot**

### CODE REVIEW

File: commons/src/main/java/org/sonarsource/analyzer/commons/ShannonEntropy.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    if (str == null || str.isEmpty()) {
+      return 0.0d;
+    }
+    int length = str.length();
+    return str.chars()
+      .collect(HashMap<Integer, Integer>::new, (map, ch) -> map.merge(ch, 1, Integer::sum), HashMap::putAll)
+      .values().stream()
+      .mapToDouble(count -> ((double) count) / length)
+      .map(frequency -> -frequency * Math.log(frequency))
+      .sum() / LOG_2;
```
Comment (ghislainpiot) [inline commons/src/main/java/org/sonarsource/analyzer/commons/ShannonEntropy.java:42]: I moved the division from the `map` to outside of it, it does not seem to affect numerical stability in the few tests

File: commons/src/main/java/org/sonarsource/analyzer/commons/ShannonEntropy.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    // utility class
+  }
+
+  public static double calculate(@Nullable String str) {
+    if (str == null || str.isEmpty()) {
+      return 0.0d;
+    }
+    int length = str.length();
+    return str.chars()
+      .collect(HashMap<Integer, Integer>::new, (map, ch) -> map.merge(ch, 1, Integer::sum), HashMap::putAll)
```
Comment (maksim-grebeniuk-sonarsource) [inline commons/src/main/java/org/sonarsource/analyzer/commons/ShannonEntropy.java:38]: could be simplifies:
```java
return str.chars()
        .boxed()
        .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()))
        .values()
        .stream()
        .map(Long::doubleValue)
        .mapToDouble(count -> count / length)
        .map(frequency -> -frequency * Math.log(frequency))
        .sum() / LOG_2;
```

**Review by maksim-grebeniuk-sonarsource — APPROVED** [review]
LGTM, just one super minor suggestion to simplify the stream API used

File: commons/src/main/java/org/sonarsource/analyzer/commons/ShannonEntropy.java (hunk 2 above)
Comment (ghislainpiot) [reply commons/src/main/java/org/sonarsource/analyzer/commons/ShannonEntropy.java:38]: I was reusing the previous SonarJava implementation, but if this one also works I can change it.

**Pull request merged**